lib/block-supports/custom-css.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Custom CSS block support.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Stores the custom CSS for block instances to be output in the footer.
+ *
+ * @var string
+ */
+global $gutenberg_block_custom_css;
+$gutenberg_block_custom_css = '';
+
+/**
+ * Get the custom CSS class name for a block.
+ *
+ * @since 7.0.0
+ *
+ * @param array $block Block object.
+ * @return string The custom CSS class name.
+ */
+function gutenberg_get_custom_css_class_name( $block ) {
+	return wp_unique_id_from_values( $block, 'wp-custom-css-' );
+}
+
+/**
+ * Render the custom CSS stylesheet and add class name to block as required.
+ *
+ * @since 7.0.0
+ *
+ * @param array $parsed_block The parsed block.
+ * @return array The same parsed block with custom CSS class name added if appropriate.
+ */
+function gutenberg_render_custom_css_support_styles( $parsed_block ) {
+	global $gutenberg_block_custom_css;
+
+	$custom_css = $parsed_block['attrs']['style']['css'] ?? null;
+
+	if ( empty( $custom_css ) ) {
+		return $parsed_block;
+	}
+
+	// Generate a unique class name for this block instance.
+	$class_name         = gutenberg_get_custom_css_class_name( $parsed_block );
+	$updated_class_name = isset( $parsed_block['attrs']['className'] )
+		? $parsed_block['attrs']['className'] . " $class_name"
+		: $class_name;
+
+	_wp_array_set( $parsed_block, array( 'attrs', 'className' ), $updated_class_name );
+
+	// Process the custom CSS using the same method as global styles.
+	$selector      = '.' . $class_name;
+	$processed_css = WP_Theme_JSON_Gutenberg::process_blocks_custom_css( $custom_css, $selector );
+
+	if ( ! empty( $processed_css ) ) {
+		// Store the CSS to be output later.
+		$gutenberg_block_custom_css .= $processed_css;
+	}
+
+	return $parsed_block;
+}
+
+/**
+ * Outputs the collected block custom CSS in the footer.
+ *
+ * The CSS is output in the footer because it's collected during block rendering
+ * (via render_block_data filter), which happens after wp_head.
+ *
+ * @since 7.0.0
+ */
+function gutenberg_output_block_custom_css() {
+	global $gutenberg_block_custom_css;
+
+	if ( empty( $gutenberg_block_custom_css ) ) {
+		return;
+	}
+
+	echo '<style id="wp-block-custom-css">' . $gutenberg_block_custom_css . '</style>';
+}
```

Comment:
The custom styles output seems odd to me. I understand this is meant to ensure that custom CSS loads last, but we could use the existing style-loading system to do that.

@westonruter, you did some similar work in WP 6.9. Do you have any suggestions for improving it?

P.S. Also not a big fan of the new global variable 😅